Various fixes and update of ResultWriter

HomesGH opened the pull request:
# Description

To following was done:

1. Fixes in 

## How Has This Been Tested?

- Comparison of output of ResultWriter with output of RegionSampling plugin

amartyads commented on src/io/ResultWriter.cpp:
```diff
@@ -1,89 +1,111 @@
+#ifdef ENABLE_MPI
+    MPI_Allreduce(&uKinLocal, &uKinGlobal, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
+	MPI_Allreduce(&uKinTransLocal, &uKinTransGlobal, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
+	MPI_Allreduce(&uKinRotLocal, &uKinRotGlobal, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
```
Can we use `DomainDecompMPIBase.h::getCommunicator()` here instead of `MPI_COMM_WORLD`? In my particular case I'm saying this because when I have multiple MD instances running with MaMiCo, each of the instances have separate communicators. But generally, won't the code be better if we use the communicator? Or is it unnecessary for result writing anyway?

HomesGH replied:
To be honest, I am not exactly sure what the difference is. But I think, that it does not really matter for the ResultWriter. I changed it to `DomainDecompMPIBase.h::getCommunicator()`.

FG-TUM commented on src/molecules/Site.h:
```diff
@@ -126,7 +127,15 @@ class LJcenter : public Site {
+		// Read shifted value and check if its boolean (true/false) to avoid legacy-errors
+		std::string strShifted = "false";
+		xmlconfig.getNodeValue("shifted", strShifted);
+		if (strShifted == "true" || strShifted == "false") {
```
```suggestion
        std::transform(strShifted.begin(), strShifted.end(), strShifted.begin(), [](auto c) { return std::tolower(c); });
		if (strShifted == "true" || strShifted == "false") {
```
Force the input to lowercase before comparing to avoid annoying input errors.
(Might need an `#include <algorithm>`)

FG-TUM commented on src/io/ResultWriter.cpp:
```diff
@@ -1,89 +1,111 @@
-			<< std::setw(_writePrecision+15) << "beta_trans"
-			<< std::setw(_writePrecision+15) << "beta_rot"
 			<< std::setw(_writePrecision+15) << "c_v"
 			<< std::setw(_writePrecision+15) << "N"
```
Why not create a second member like `_writeWidth = _writePrecision + 15`? Makes changes easier.

HomesGH commented on src/io/ResultWriter.cpp:
```diff
@@ -1,89 +1,111 @@
+				<< std::setw(_writePrecision+15) << std::scientific << std::setprecision(_writePrecision) << _p_acc/_numSamples
+				<< std::setw(_writePrecision+15) << std::scientific << std::setprecision(_writePrecision) << domain->cv()
+				<< std::setw(_writePrecision+15) << std::scientific << std::setprecision(_writePrecision) << globalNumMolecules
+				<< endl;
```
I introduced a lambda function, see [here](https://github.com/ls1mardyn/ls1-mardyn/blob/0d151cb9212fb29fc87e7a94bb3d24f004ca6687/src/io/ResultWriter.cpp#L92). Please have a look if it seems to be what you want as I am not so familiar with lambda functions ;-)

amartyads commented on src/io/ResultWriter.cpp:
```diff
@@ -1,89 +1,111 @@
-			<< std::setw(_writePrecision+15) << "beta_rot"
 			<< std::setw(_writePrecision+15) << "c_v"
 			<< std::setw(_writePrecision+15) << "N"
 			<< endl;
```
I created #259 for the namespace issue. Maybe the neater solution is to use a `using namespace std;` only inside the functions where it's needed, too scope it properly?

FG-TUM replied:
The namespaceing issue will be tackled in a dedicated PR.

amartyads commented on src/molecules/Site.h:
```diff
@@ -126,7 +128,17 @@ class LJcenter : public Site {
+		// Convert to lower case to avoid input errors
+		std::transform(strShifted.begin(), strShifted.end(), strShifted.begin(), [](auto c) { return std::tolower(c); });
+		if (strShifted == "true" || strShifted == "false") {
+			xmlconfig.getNodeValue("shifted", _shiftRequested);
```
```suggestion
			_shiftRequested = (strShifted == "true") ? true : false;
```

HomesGH replied:
Fixed, see [commit ](https://github.com/ls1mardyn/ls1-mardyn/pull/255/commits/be04e8b88298bb0eaf9e4ceab5fc6f2be9ca040d).

cniethammer commented on src/molecules/Site.h:
```diff
@@ -126,7 +130,17 @@ class LJcenter : public Site {
 		xmlconfig.getNodeValueReduced("sigma", _sigma);
-		xmlconfig.getNodeValue("shifted", _shiftRequested);
+		// Read shifted value and check if its boolean (true/false) to avoid legacy-errors
+		std::string strShifted = "false";
+		xmlconfig.getNodeValue("shifted", strShifted);
+		// Convert to lower case to avoid input errors
+		std::transform(strShifted.begin(), strShifted.end(), strShifted.begin(), [](auto c) { return std::tolower(c); });
+		if (strShifted == "true" || strShifted == "false") {
+			std::istringstream(strShifted) >> std::boolalpha >> _shiftRequested;
+		} else {
+			global_log->error() << "Parameter <shifted> of components has to be either set to 'true' or 'false'" << std::endl;
+			mardyn_exit(1);
+		}
 	}
```
This should not be necessary at as the xml parser code already handles "YES", "TRUE", "ON" ignoring case. See https://github.com/ls1mardyn/ls1-mardyn/blob/b6f274e625fe7a8c07861f09c2b6c9c5e0243f2e/src/utils/xmlfile.cpp#L589

HomesGH replied:
The point of this code is to check if the input value of `<shifted>` is really a boolean as it was a float before introducing the `_shiftRequested` variable a few month ago. If we would just do `xmlconfig.getNodeValue("shifted", _shiftRequested);` then a float will be probably converted to false. Even though a user would expect a number != 0 to shift the potential. This code makes sure that an old version of the components where the shift value was set to be `0.097901347` (shift for cutoff radius =2.5) can not be used without an error. If we just parsed this value, the potential would not be shifted and the simulation run would yield wrong results.

HomesGH replied:
This is [now](https://github.com/ls1mardyn/ls1-mardyn/pull/255/commits/981cddfbf729484d9b2e30adb4019de07b166076) solved by checking all boolean input values, and [additionally](https://github.com/ls1mardyn/ls1-mardyn/pull/255/commits/0f8092e6329dc537d166008ec378c6ad7af36709) also other user input via the xmlfile.

cniethammer commented on src/utils/Logger.cpp:
```diff
@@ -29,6 +31,7 @@ Logger::Logger(logLevel level, std::string prefix) :  _log_level(level), _msg_lo
 	_filename = filenamestream.str();
 	_log_stream = new std::ofstream(_filename.c_str());
+	*_log_stream << std::boolalpha;  // Print boolean as true/false
 }
```
Changes to the logger should go to a separate commit/PR IMHO.

HomesGH replied:
Maybe for now, it is a bit pointless to move it to a separate commit as I would have to delete it in a first commit and add it again in a second one. A PR would be a bit over the top for one line...
But in general, you are right, I will consider it next time.

HomesGH commented on src/io/ResultWriter.cpp:
```diff
@@ -1,89 +1,116 @@
-		const auto nowStr = std::put_time(localtime_r(&now, &unused), "%c");
-		_resultStream << "# ls1 mardyn simulation finished at " << nowStr << endl;
-		_resultStream.close();
+	if ((simstep % _writeFrequency == 0) and (simstep > 0UL)) {
+		// Only main rank writes data to file
+		if (domainDecomp->getRank() == 0) {
+			const string resultfile(_outputPrefix+".res");
+			std::ofstream resultStream;
+			resultStream.open(resultfile.c_str(), std::ios::app);
+			auto printOutput = [&](auto value) {
+				resultStream << std::setw(_writeWidth) << std::scientific << std::setprecision(_writePrecision) << value;
+			};
+			resultStream << std::setw(10) << simstep;
+				printOutput(_simulation.getSimulationTime());
+				printOutput(_uPot_acc/_numSamples);
+				printOutput(_uKin_acc/_numSamples);
+				printOutput(_uKinTrans_acc/_numSamples);
+				printOutput(_uKinRot_acc/_numSamples);
+				printOutput(_p_acc/_numSamples);
+				printOutput(domain->cv());
+				printOutput(globalNumMolecules);
+				resultStream << std::endl;
+			resultStream.close();
+		}
+
+		// Reset values
+		_numSamples = 0UL;
+		_uPot_acc = 0.0F;
+		_uKin_acc = 0.0F;
+		_uKinTrans_acc = 0.0F;
+		_uKinRot_acc = 0.0F;
+		_p_acc = 0.0F;
```
The existing implementation was not what was [documented](https://github.com/ls1mardyn/ls1-mardyn/blob/b6f274e625fe7a8c07861f09c2b6c9c5e0243f2e/src/io/ResultWriter.h#LL13C45-L13C45). For me, it seems like the accumulator returns not only the value averaged over the last `_writeFrequency` steps but for the whole simulation. Otherwise it would had to reset the data after getting the average which it does not.
What do you mean by "higher write frequency than the sample count"? The sample count just counts the number of sampling steps. It is often the writing frequency but not at the very beginning (simstep=0 is sampled but it would lead to wrong averages if at e.g. simstep=10=writefrequency) the accumulated value is divided by 10 and not 11.)
I not that it is not perfect to have the loop and MPI call in here. But at the moment, the domain/ensemble does not provide a working method to get the global kinetic energy.

HomesGH replied:
@cniethammer So what do you suggest regarding my 3 comments?

HomesGH commented:
@FG-TUM 
Please have a look at point 7 and 8 in the description of this PR and commit [845b160](https://github.com/ls1mardyn/ls1-mardyn/pull/255/commits/845b160608485d5dda8cf9ee3155fa07033fd59b). Especially point 8 as it is about AutoPas.

HomesGH commented:
@cniethammer 
There is something buggy in the MMPLDWriter.cpp file. I narrowed the issue down to the [MPI_file_seek/MPI_get_position](https://github.com/ls1mardyn/ls1-mardyn/blob/d4e5c8f2c0dfbacb84459db33e7d1e952cc1d645/src/io/MmpldWriter.cpp#L330-L332) in the `finish` method. I kind of fixed it with commit [d4e5c8f](https://github.com/ls1mardyn/ls1-mardyn/pull/255/commits/d4e5c8f2c0dfbacb84459db33e7d1e952cc1d645). The problem is, that `endPosition` is at max `4194304`, which seems to be a block size (?). This is strange and as you know MPI much better, your help is much appreciated.

I was also running a test with a different number of frames. When endPosition=_seekTable.back(), the file was not corrupted according to MMPLD tool [mmpldinfo.py](https://github.com/UniStuttgart-VISUS/megamol/blob/master/utils/MMPLD/mmpldinfo.py). 
![table_MMPLDTest](https://github.com/ls1mardyn/ls1-mardyn/assets/55833544/ee7c28c5-d556-4d6e-bdf5-2828d96d458b)

amartyads commented on src/Domain.h:
```diff
@@ -415,6 +422,10 @@ class Domain {
 	double _globalUpotCspecif;
+	//! global two times translational kinetic energy
+	double _globalsummv2;
+	//! global two times rotational kinetic energy
```
```suggestion
	//! global rotational kinetic energy times two
```

HomesGH commented on src/io/ResultWriter.cpp:
```diff
@@ -1,88 +1,86 @@
 
-void ResultWriter::finish(ParticleContainer * /*particleContainer*/,
-						  DomainDecompBase *domainDecomp, Domain * /*domain*/){
+	if ((simstep % _writeFrequency == 0) and (simstep > 0UL)) {
```
Yes, in theory, you are right. But as far as I know, this is almost nowhere done.
In addition, you can also just enter a negative number and you will never notice, as it will be converted to a (really huge) unsigned one. This could be solved by using a signed integer for `_writeFrequency` and checking for `<=0` but it would be better to implement those parsing checks more fundamentally. The whole input parsing/sanity check does not really exist (only for boolean now).
A possible solution could start [here](https://github.com/ls1mardyn/ls1-mardyn/blob/981cddfbf729484d9b2e30adb4019de07b166076/src/utils/xmlfile.cpp#L531-L608); maybe by adding more explicit functions (and checks).

amartyads replied:
you're right, we should have more comprehensive input checks. It's just that normally a frequency 0 in any writing function should mean that nothing should be written, so it's not unreasonable for someone to assume that 0 is valid input. I think the check you've included is fine here.

FG-TUM reviewed (approved):
Although I agree with @cniethammer's comment in principle, I think that the solution proposed in this PR is "good enough" for now so that we can move on. But I'd like to hear @cniethammer's opinion.

HomesGH commented:
> Although I agree with @cniethammer's comment in principle, I think that the solution proposed in this PR is "good enough" for now so that we can move on. But I'd like to hear @cniethammer's opinion.

Do you mean the comment on the xml parser code for boolean? I revoked the special check for `<shifted>` and implemented the [parsing of boolean](https://github.com/ls1mardyn/ls1-mardyn/pull/255/files#diff-239732a8307f579f63d703f7caaa825b12e4c6121fd3813ddac5e0dc237703e3R572) properly instead. In addition, I just [added some basic checks](https://github.com/ls1mardyn/ls1-mardyn/pull/255/commits/0f8092e6329dc537d166008ec378c6ad7af36709) also for the template parsing function and deleted the specified `getValue()` methods for int, long, float and double. Those do not cover all cases (e.g. unsigned int) anyway.

FG-TUM commented:
No I was referring to this comment:
>Also this moves time consuming MPI collectives into the plugin - if all plugins compute energies by themselves this will lead to unnecessary overheads - where the domain/ensemble classes will compute once and cache the values.

Not sure what functionality is lost though...

HomesGH commented:
> No I was referring to this comment:
> 
> > Also this moves time consuming MPI collectives into the plugin - if all plugins compute energies by themselves this will lead to unnecessary overheads - where the domain/ensemble classes will compute once and cache the values.
> 
> Not sure what functionality is lost though...

Ok, I see. In a former version of this PR, I calculated the kinetic energies in the plugin itself. But this is now done in the Domain class since it was calculated there anyway for the thermostat (but not stored yet).
I am also not sure regarding the lost functionality. In my opinion, it does not make sense to write results more often than you sample. This would result in duplicated values, wouldn't it?

I would also be much interested in the issue regarding the MMPLDWriter. This bug was very strange for me :D

cniethammer commented:
I now had some time to look at this. Honestly, this seems an agglomerate of changes that touch various parts besides the ResultWriter - including unrelated fixes in e.g. the Molecule class or MmpldWriter. IMHO, this should be split into several clean PRs that address single aspects.

HomesGH commented:
> I now had some time to look at this. Honestly, this seems an agglomerate of changes that touch various parts besides the ResultWriter - including unrelated fixes in e.g. the Molecule class or MmpldWriter. IMHO, this should be split into several clean PRs that address single aspects.

@cniethammer 
Yes, it is not ideal. But this is due to the fact that the review here takes a long time and ls1 is full of bugs.
I, personally, do not have the time besides my research to split this PR up, just to make it a bit cleaner. The important things are the single commits and that there are no different issues addressed in one commit.
So either somebody else "cleans" this PR or we can take it as it is.

cniethammer commented:
OK, I will clean this PR up by splitting it into separate PRs.

HomesGH commented:
> OK, I will clean this PR up by splitting it into separate PRs.

Alright, thank you :)

HomesGH commented:
@cniethammer 
Can we please continue with the work on this PR or the splitted ones? I really need those fixes and improvements for my work. Especially the ones with the MMPLDWriter and the xml input parsing would be good.

HomesGH commented:
> OK, I will clean this PR up by splitting it into separate PRs.

Hello @cniethammer ,
do you have any time to continue with the merging/creation of separate PRs?
